state-chain/pallets/cf-ingress-egress/src/lib.rs

```diff
@@ -1251,6 +1327,26 @@ impl<T: Config<I>, I: 'static> IngressSink for Pallet<T, I> {
 }
 
 impl<T: Config<I>, I: 'static> Pallet<T, I> {
+	fn mark_transaction_as_tainted_inner(
+		origin: OriginFor<T>,
+		tx_id: <T::TargetChain as Chain>::DepositDetails,
+	) -> DispatchResult {
+		let account_id = ensure_signed(origin)?;
+		let is_broker = T::AccountRoleRegistry::has_account_role(&account_id, AccountRole::Broker);
+		let is_lp =
+			T::AccountRoleRegistry::has_account_role(&account_id, AccountRole::LiquidityProvider);
+		ensure!(is_broker || is_lp, BadOrigin);
```

Comment:
Actually I'm not sure we want/need LPs to mark transactions as tainted. We need their deposit address *in case* a tx is marked, but in general they won't be marking their own.

We will probably add another account role or similar for the 'reporter' role, but that can be in a future PR. For now we can remove the LP here.

Reply:
I'd also suggest unwrapping the origin and doing permission checks in the extrinsic rather than in this `inner` method. Might make testing/benchmarking easier.

Reply:
"Inner" was a more temporary solution since we only wanted to support BTC for now and that would open a whole new number of problems with tests and benchmarks. Since I think it could make sense to support at least ETH as well, we can do this properly and the inner function warp can disappear. In general, I think we should try to benchmark as close as possible to be on the safe side.